It was proposed by Christian Goldbach that every odd composite number can be written as the sum of a prime and twice a square.

$$
9 = 7 + 2×1^2 \\
15 = 7 + 2×2^2 \\
21 = 3 + 2×3^2 \\
25 = 7 + 2×3^2 \\
27 = 19 + 2×2^2 \\
33 = 31 + 2×1^2
$$

It turns out that the conjecture was false.

What is the smallest odd composite that cannot be written as the sum of a prime and twice a square?

### Version 1: Brute force

Formally stated, Goldbach's other conjectures says that for all odd composite numbers $n$ can be expressed as $n = 2k^2 + p$ for some integer $k$ and prime number $p$. Let $S_n = \{ n - 2k^2 : k = 1, 2, \dotsc, \lfloor \sqrt{\frac{n}{2}} \rfloor \}$. If any element $n - 2k^2$ of $S_n$ is prime, then we call $k$ a *witness* to Goldbach's other conjecture. Let $P_n = \{p : p \text{ is prime and } p < n\}$. Then our algorithm searches for the smallest $n$ such that $P_n \cap S_n = \emptyset$, providing a counterexample to Goldbach's other conjecture.

In [1]:
from itertools import count
from collections import defaultdict

We augment a very space-efficient implementation (due to David Eppstein, UC Irvine) of the Sieve of Erastothenes to generate *composite* numbers and keep track of all prime numbers below it. In the outermost-loop below, `primes` will always be the list of all prime number strictly below $n$, i.e. it is equivalent to $P_n$. Note that if $n$ is odd and composite, then the largest prime below it is no greater than $n-2$, since $n-1$ is even. Since $\max S_n = n - 2$ searching through $P_n$ is sufficient (i.e. if $n-2$ is prime, then $n-2 \in P_n$.) The loop terminates when we encounter an $n$ with no witnesses.

In [2]:
factors = defaultdict(list)
witness = {}
primes = []
for n in count(2):
    if factors[n]:
        # n is composite
        for m in factors.pop(n):
            factors[n+m].append(m)
        if n % 2:
            # n is odd and composite
            for k in xrange(1, int((n/2)**0.5)+1):
                p = n - 2*k**2
                if p in primes: 
                    # TODO: `in` is O(len(primes))
                    # could optimize by using set
                    witness[n] = k
                    break
            if not n in witness:
                break
    else:
        factors[n*n].append(n)
        primes.append(n)

In [3]:
n

5777

The answer is 5777.

Note that not only is this implementation space-efficient, it is also very time efficient, since we incrementally build our list of primes, composites and witnesses incrementally from bottom-up. If we hadn't augmented the implementation of the prime sieve, we would have had to use the prime sieve to obtain all odd composites, and perform a primality test on all elements of $S_n$, which would (usually) require a prime factorization algorithm, which in turn (usually) requires a prime sieve, not to mention the fact that there would be many overlapping subproblems, i.e. many $n < m$ such that $S_n \cap S_m \ne \emptyset$, so we'd have to use dynamic programming by for example memoizing the primality testing function or some other optimization - just a whole mess of redundancies and inefficiencies.

In [4]:
witness

{9: 1,
 15: 1,
 21: 1,
 25: 1,
 27: 2,
 33: 1,
 35: 3,
 39: 1,
 45: 1,
 49: 1,
 51: 2,
 55: 1,
 57: 5,
 63: 1,
 65: 3,
 69: 1,
 75: 1,
 77: 3,
 81: 1,
 85: 1,
 87: 2,
 91: 1,
 93: 4,
 95: 6,
 99: 1,
 105: 1,
 111: 1,
 115: 1,
 117: 2,
 119: 3,
 121: 2,
 123: 5,
 125: 3,
 129: 1,
 133: 1,
 135: 2,
 141: 1,
 143: 6,
 145: 2,
 147: 2,
 153: 1,
 155: 3,
 159: 1,
 161: 6,
 165: 1,
 169: 1,
 171: 2,
 175: 1,
 177: 5,
 183: 1,
 185: 3,
 187: 2,
 189: 2,
 195: 1,
 201: 1,
 203: 6,
 205: 2,
 207: 2,
 209: 3,
 213: 1,
 215: 3,
 217: 3,
 219: 2,
 221: 6,
 225: 1,
 231: 1,
 235: 1,
 237: 2,
 243: 1,
 245: 3,
 247: 2,
 249: 2,
 253: 1,
 255: 4,
 259: 1,
 261: 4,
 265: 1,
 267: 8,
 273: 1,
 275: 3,
 279: 1,
 285: 1,
 287: 3,
 289: 2,
 291: 2,
 295: 1,
 297: 7,
 299: 3,
 301: 2,
 303: 4,
 305: 6,
 309: 1,
 315: 1,
 319: 1,
 321: 2,
 323: 6,
 325: 2,
 327: 5,
 329: 3,
 333: 1,
 335: 3,
 339: 1,
 341: 6,
 343: 4,
 345: 2,
 351: 1,
 355: 1,
 357: 2,
 361: 1,
 363: 4,
 365: 3,
 369: 1,
 371: 3,
 375: 1,


Let's list out the witnesses to all numbers below 5777.

In [5]:
for n in witness:
    print '{0} = 2 {1}^2 + {2}'.format(n, witness[n], n - 2*witness[n]**2)

9 = 2 1^2 + 7
15 = 2 1^2 + 13
21 = 2 1^2 + 19
25 = 2 1^2 + 23
27 = 2 2^2 + 19
33 = 2 1^2 + 31
35 = 2 3^2 + 17
39 = 2 1^2 + 37
45 = 2 1^2 + 43
49 = 2 1^2 + 47
51 = 2 2^2 + 43
55 = 2 1^2 + 53
57 = 2 5^2 + 7
63 = 2 1^2 + 61
65 = 2 3^2 + 47
69 = 2 1^2 + 67
75 = 2 1^2 + 73
77 = 2 3^2 + 59
81 = 2 1^2 + 79
85 = 2 1^2 + 83
87 = 2 2^2 + 79
91 = 2 1^2 + 89
93 = 2 4^2 + 61
95 = 2 6^2 + 23
99 = 2 1^2 + 97
105 = 2 1^2 + 103
111 = 2 1^2 + 109
115 = 2 1^2 + 113
117 = 2 2^2 + 109
119 = 2 3^2 + 101
121 = 2 2^2 + 113
123 = 2 5^2 + 73
125 = 2 3^2 + 107
129 = 2 1^2 + 127
133 = 2 1^2 + 131
135 = 2 2^2 + 127
141 = 2 1^2 + 139
143 = 2 6^2 + 71
145 = 2 2^2 + 137
147 = 2 2^2 + 139
153 = 2 1^2 + 151
155 = 2 3^2 + 137
159 = 2 1^2 + 157
161 = 2 6^2 + 89
165 = 2 1^2 + 163
169 = 2 1^2 + 167
171 = 2 2^2 + 163
175 = 2 1^2 + 173
177 = 2 5^2 + 127
183 = 2 1^2 + 181
185 = 2 3^2 + 167
187 = 2 2^2 + 179
189 = 2 2^2 + 181
195 = 2 1^2 + 193
201 = 2 1^2 + 199
203 = 2 6^2 + 131
205 = 2 2^2 + 197
207 = 2 2^2 + 199
209 = 2 3^2 